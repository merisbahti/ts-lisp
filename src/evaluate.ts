import * as T from './types'
import { Result, Val, Err, fmap } from './result'
import { Env, StdLib, lookUp } from './environment'
import { pp } from './prettyPrint'

export const evaluate = (e: T.Expr, env: Env = StdLib): Result<T.Expr> => {
  switch (e.type) {
    case 'atom':
      return fmap(lookUp(e.v, env), (e) => evaluate(e, env))
    case 'number':
      return Val(e)
    case 'procedure':
      return Err('Cannot evaluate procedure')
    case 'list':
      const head = e.v[0]
      if (!head) {
        return Err('Cannot evaluate empty list.')
      }
      if (head.type !== 'atom') {
        return Err(`Cannot evaluate using ${pp(head)}.`)
      }
      return fmap(
        lookUp(head.v, env),
        (lookedUp: T.Expr) => {
          if (lookedUp.type === 'procedure') {
            return lookedUp.v(...e.v.slice(1))
          } else {
            return Err(`Cannote evaluate non-procedure, found: ${pp(head)}`)
          }
        }
      )
  }
}
