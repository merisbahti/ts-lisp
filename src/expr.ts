import * as T from './types'

export const Num = (v: number): T.Expr => ({ type: 'number', v })
export const List = (...v: Array<T.Expr>): T.Expr => ({ type: 'list', v })
export const Atom = (v: string): T.Expr => ({ type: 'atom', v })
export const Proc = (v: T.ProcValue): T.Expr => ({ type: 'procedure' as 'procedure', v })
