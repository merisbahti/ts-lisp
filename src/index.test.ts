import { Num, List, Atom } from './expr'
import {Â evaluate } from './evaluate'

test('simple addition', () => {
  expect(
    evaluate(
      List(
        Atom('add'),
        Num(5),
        Num(3)))
  ).toMatchSnapshot()
  expect(
    evaluate(
      List(
        Atom('add'),
        Num(5),
        Num(-3)))
  ).toMatchSnapshot()
})
test('failure cases', () => {
  expect(
    evaluate(
      List(
        Atom('add'),
        Atom('hey'),
        Num(-3)))
  ).toMatchSnapshot()
  expect(
    evaluate(
      List(
        Atom('add'),
        Num(-3)))
  ).toMatchSnapshot()
  expect(
    evaluate(
      List(
        Atom('add'),
        Atom('foo'),
        Num(3),
        Num(-3)))
  ).toMatchSnapshot()
})
