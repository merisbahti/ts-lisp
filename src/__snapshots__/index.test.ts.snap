// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failure cases 1`] = `
Object {
  "error": "hey could not be found in the env.",
  "t": "err",
}
`;

exports[`failure cases 2`] = `
Object {
  "error": "'add' requires two arguments, recieved values: -3",
  "t": "err",
}
`;

exports[`failure cases 3`] = `
Object {
  "error": "'add' requires two arguments, recieved values: foo,3,-3",
  "t": "err",
}
`;

exports[`simple addition 1`] = `
Object {
  "t": "value",
  "value": Object {
    "type": "number",
    "v": 8,
  },
}
`;

exports[`simple addition 2`] = `
Object {
  "t": "value",
  "value": Object {
    "type": "number",
    "v": 2,
  },
}
`;
